### ISSUE TITLE
Core language syntax for constructors

### ISSUE THREAD

**Issue by EliasC** [miking-lang/miking#31]
Since we are intending for MExpr to be a core language, there are some simplifications that could still be made. This issue is for discussing the syntax of constructors from this angle.

At the moment, the syntax for function application and constructor application is overloaded:

```
con Foo : Int -> T in
let f = lam x. x in
(Foo 42, f 42)
```

This makes perfect sense from a usability point of view (the OCaml developers might disagree), but it introduces some overhead. Currently, it is a dynamic overhead; when evaluating an application, we branch on what kind of value the left-hand side is. With a type system it would be a static overhead instead, since we have to analyze the program to figure out if the left-hand side is a constructor or a function. Even if this doesn't affect run- or compile times significantly, it still adds complexity.

In order to remove this complexity, we would need a _syntactic_ distinguisher (note that we soon won't require constructors to be capitalized or identifiers to be lower case, so we cannot use that). There are two solutions which are immediately obvious:

1. Introduce a symbol that means "constructor application". For example, we could write `Foo#42`. In order to keep syntax as regular as possible, we would use this for pattern matching as well: `match x with Foo#y then ...`.
2. Let constructor application be notated using parenthesis _immediately_ after an identifier. This would mean that `Foo(42)` is constructing a `Foo`, whereas `Foo (42)` is applying the function `Foo` to `42`. By a similar argument as above, pattern matching should be adapted: `match x with Foo(y) then ...` is OK, while `match x with Foo y then ...` is not.

In both of these solutions we would also remove the possibility of using a constructor as a function directly, i.e., instead of `map Foo seq` you would write `map (lam x. Foo x) seq`. 

Again, I want to remind you that the point of this is keep the core language as simple as possible! It is obvious this is a step back usability-wise, but it makes the implementation simpler, and moves us closer to a situation where the concrete syntax is a direct representation of the abstract syntax.

**Comment by EliasC** [comment]
Personally, I feel like the first solution is uglier, but also less ad-hoc and error-prone than the second solution. I feel like having a special symbol means that I can tell at a glance if something is a constructor application or a function application, which is valuable.

**Comment by dlunde** [comment]
I do not like that "Foo(42)" and "Foo (42)" have different semantics, so I guess that makes me prefer 1. Isn't it easiest to have different types of identifiers for constructors and functions (i.e. capitalize vs not capitalize)? Or is that what we are trying to avoid? I'm not up to date on the discussion in Slack.

**Comment by johnwikman** [comment]
I also prefer the first solution, though I do not find the `match x with Foo#y then ...` syntax very readable. I would prefer something closer to this:
```
match x with Foo::y then ...
```
Alternatively something like `match x with Foo {y} then ...`, though it could be good to reserve curly brackets for sets, were it to be added in the future.

**Comment by EliasC** [comment]
> @dlunde: Isn't it easiest to have different types of identifiers for constructors and functions?

I agree that this would be simpler, but the argument against it is that it brings an unnecessary restriction which might make compilation into MExpr less straightforward if the source language does not have the same naming requirements (for example it might require some name mangling instead of just using the names as is). I think an important thing to remember is that MExpr in the end is not meant as a language to program in, even if that is what we are doing right now.

> @johnwikman: I would prefer something closer to this...

Yes, the `#` was just an arbitrary symbol as example. It's hard to find something that isn't overloaded in other languages though!

- `Foo::y`: List cons in ML-languages
- `Foo:y`: Type specifications; List cons in Haskell
- `Foo (y)`: Function application
- `Foo [y]`: Sequence literals
- `Foo {y}`: Record creation
- Anything that is normally used as an infix operator looks weird to me: `Foo&y`, `Foo/y`, `Foo*y`...

Looking at my numbers row on the keyboard, the following alternatives all look viable to me:
`Foo#y`, `Foo~y`, `Foo@y`, `Foo%y`, `Foo^y` (even though the last two are used as infix operators). Because we don't have infix comparisons I would also consider `Foo<y>` (trying to look like the unicode `Foo⟨y⟩`).

**Comment by EliasC** [comment]
Another design issue to consider is if we want to use the same syntax for type constructor application or not (picking another arbitrary symbol from above):

```
let x : Option~Int = Some~42
```

On the one hand it's very regular and it makes sense that constructing a value mirrors constructing a type, but on the other hand it might make sense to tell from a glance which one of them it is (although they can never appear in the same position). 

Instinctively I feel like they _should_ be the same.

**Comment by johnwikman** [comment]
I like `Foo<y>`, much easier to read. I am not a big fan of only using an infix symbol.

**Comment by david-broman** [comment]
After the discussion Elias and I had yesterday that resulted in this thread, I think that a third alternative might be a better choice. 

The idea is that we keep the same syntax as we have today, that is, we use juxtaposition for both function calls and construction of data constructors. I.e. there is no special space syntax (as in alternative 2 above) or special infix syntax symbol (as using # in alternative 1 above). Instead, we add a static translation step using a name environment that translates from the overloaded form, into an AST where function application and data construction are different terms. Still, we enforce that constructors are unary (which is better than what we have today). The good thing with this is also that you will get an error message if you just write `Foo` instead of `Foo()` or `Foo ()`. I think that this translation is quite simple and it does not need a complete type system. Moreover, we do not get any ad-hoc behaviour, and the syntax is actually readable. 

What do you think?

**Comment by elegios** [comment]
I don't have super strong opinion on this, other than a dislike for lexical hacks, so not the significant space option. Using a lexical analysis to distinguish function application from constructor application seems fine to me, though after discussion yesterday I think I have a weak preference for having some explicit syntactic marker. 

It is worth mentioning that the lexical analysis approach precludes that `parse . prettyprint` is the identity function, since we can construct an ast that does constructor application where the constructor is lexically shadowed by some other value (assuming some more direct method of constructing an ast than parsing). 

An additional alternative is something like `{Foo arg}`, which avoids precedence considerations and can be distinguished from records (should we choose to add them) by having a space between the two things instead of `=` or `:`.

**Comment by EliasC** [comment]
To me it seems like adding an additional pass to the compiler is more complicated than using special syntax, and it moves us further from having the concrete syntax be a direct representation of the abstract syntax. I think enforcing unary constructors is a bigger hit usability-wise (but one still I'm willing to endure) than explicating constructor application syntactically. 

Of course it will be slightly more ergonomic to write programs if there is less syntactic clutter, but who are we catering to by choosing ergonomics over simplicity? Isn't the end-goal that no one should be writing MExpr directly?

**Comment by david-broman** [comment]
@EliasC Correct, most people should in the end not write MExpr-programs, but many compiler writers will read MExpr programs. Hence, a clear syntax is good to have, if we can.
I am thinking in the lines that we had before, differentiating between MCore and Pure MCore, where Pure MCore corresponds directly to the abstract syntax and MCore is still very close, but has less syntactic noice. Also, with both, it should be clear to the reader exactly what the underlying abstract syntax is. 

In this case, my suggestion above would be MCore, and a Pure MCore output would be using an explicit construct, such as #. We already have something similar today. For instance, a string is actually a sequence of characters. We do not force the user to always write ['H','e','l','l','o'], or to get this when he/she pretty prints an MExpr. Instead, we would of course like to write "Hello". In the same line, the syntax "Hello" would be the standard MCore syntax, and ['H','e','l','l','o'] would be the Pure MCore syntax.

**Comment by EliasC** [comment]
I buy the distinction between MCore and Pure MCore! Do we allow _writing_ Pure MCore? In other words, is there a parser for Pure MCore?

**Comment by david-broman** [comment]
OK, @EliasC, @elegios , and I had a longer discussion in front of the white board. We have now concluded on a design that I believe works.

The main idea is to have different syntactic identifiers in the AST for constructors and for variables. The general way of writing a variable identifier is as `#var"foo"`, but if the identifier is starting with a lower case letter, it is enough to just write `foo` in an MCore program. The identifier for a constructor is `#con"Foo"`, but if the identifier starts with a capital letter, it is enough to write `Foo`. Hence, we can separate between constructors and variables without introducing an extra translation step or to have extra syntactic symbols. 

Note also that there is no restriction on the identifiers. For instance `#con"foo ident"` is a constructor that includes white space and starts with a lower case character.  We will also use the same pattern for types, e.g `#type"Name"`.

I will start to implement this now, so please react to this as soon as possible.

**Comment by johnwikman** [comment]
Does this mean that I must write `#var"foo"` or can I write it as `#var "foo"`, `#var("foo")`, or `#var ("foo")` as well?

**Comment by EliasC** [comment]
@johnwikman This means that you can write `foo`, but that `#var"foo"` is equivalent (you only need this notation when you have an identifier that does not follow the naming rules). @david-broman will have the final say, but my feeling is that only exactly `#var"foo"` is going to be allowed (`#var "foo"` would be a parse error).

**Comment by david-broman** [comment]
@johnwikman and @EliasC . Yes, the idea was to do as @EliasC says. An identifier `#var"foo"` is a lexical token, not an application that is parsed. 
@johnwikman Do you have any strong opinion about this syntax?

**Comment by johnwikman** [comment]
No, I think this syntax rather makes much sense.

**Comment by david-broman** [comment]
Solved in PR #90.
